In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.7 MB 29.2 MB/s 
     |████████████████████████████████| 49 kB 5.6 MB/s 


In [2]:
import pandas as pd
import numpy as np

In [3]:
from pyomo.environ import *

## Ques1. Consider the following Optimization problem:
min $f(x_1,x_2,x_3,x_4,x_5) = (x_1-1)^2+(x_1-x_2)^2+(x_2-x_3)^2+(x_3-x_4)^4+(x_4-x_5)^4,  $ \\
st. $x_1+x_2^2+x_3^3-2-3\sqrt2=0 \\
x_2-x_3^2+x_4+2-2\sqrt2=0 \\
x_1x_5-2=0 \\
-5\leq x_3\leq5 \\
0\leq x_3 \leq 3 $ \\
Solve the above optimization problem with initail values starting point $(x_1, x_2, x_3, x_4, x_5) $= (2, 2, 2, 2, 2)
## Solution: 
Since given optimization problem is non linear optimization problem.So using pyomo:




In [4]:
# Creating a model
model=ConcreteModel()

In [5]:
model.x1 = Var(initialize=2.)
model.x2 = Var(initialize=2.)
model.x3 = Var(initialize=2.)
model.x4 = Var(initialize=2.)
model.x5 = Var(initialize=2.)

In [6]:
model.constr1= Constraint(expr= model.x1+(model.x2)**2 + (model.x3)**3 -2 -3*(2)**0.5==0)
model.constr2= Constraint(expr= model.x2-(model.x3)**2 +model.x4+ 2-2*(2)**0.5==0)
model.constr3= Constraint(expr= model.x1*model.x5-2==0)
model.constr4= Constraint(expr= model.x1>=-5)
model.constr5= Constraint(expr= model.x1<=5)
model.constr6= Constraint(expr= model.x3>=0)
model.constr7= Constraint(expr= model.x3<=3)

In [7]:
def fun(model):
  return (model.x1-1)**2 + (model.x1-model.x2)**2 + (model.x2-model.x3)**2+ (model.x3-model.x4)**4 + (model.x4-model.x5)**4

In [8]:
model.cost= Objective(rule=fun, sense=minimize)

In [9]:
model.pprint()

5 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :   2.0 :  None : False : False :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :   2.0 :  None : False : False :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :   2.0 :  None : False : False :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :   2.0 :  None : False : False :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :   2.0 :  None : False : False :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : (x1 - 1)**2 + (x1 - x2)**2 + (x2 - x3)**2 + (x3 - x4)**4 + (x4 - x5)**4

In [10]:
!wget -N -q "https://ampl.com/dl/open/ipopt/ipopt-linux64.zip"
!unzip -o -q ipopt-linux64

In [11]:
solver = SolverFactory('ipopt', executable='/content/ipopt')

In [12]:
results = solver.solve(model)

In [13]:
print(results)


Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 5
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.12.13\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.023660898208618164
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [14]:
results.solver.time

0.023660898208618164

In [15]:
print('\nCost = ', model.cost())

print('\nOptimal solution:')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())

print('\nConstraints')
print('Constraint 1  = ', model.constr1())
print('Constraint 2  = ', model.constr2())
print('Constraint 3  = ', model.constr3())
print('Constraint 4  = ', model.constr4())
print('Constraint 5  = ', model.constr5())
print('Constraint 6  = ', model.constr6())
print('Constraint 7  = ', model.constr7())




Cost =  0.07877682087105699

Optimal solution:
x1 =  1.1911274562403504
x2 =  1.3626031649434767
x3 =  1.4728179315306016
x4 =  1.6350166192407933
x5 =  1.6790814362660684

Constraints
Constraint 1  =  1.7763568394002505e-15
Constraint 2  =  -4.440892098500626e-16
Constraint 3  =  -3.774758283725532e-15
Constraint 4  =  1.1911274562403504
Constraint 5  =  1.1911274562403504
Constraint 6  =  1.4728179315306016
Constraint 7  =  1.4728179315306016


## 3.[R] Report the solver status, time taken to solve and solver termination condition.
## Answer: 
Solver status: Ok \\

Time taken to solve:  0.025565147399902344 \\
Solver Termination Condition: Optimal

## 4. [R] Report the optimal objective function value and optimal values of the optimization variables.
## Answer:
Optimal objective function value: 0.07877682087105699 \\
Optimal values of variables: \\
$x_1 =  1.1911274562403504 \\
x_2 =  1.3626031649434767 \\
x_3 =  1.4728179315306016 \\
x_4 =  1.6350166192407933 \\
x_5 =  1.6790814362660684$


## 5. [R] Report if the constraints are active or inactive at the optimal solution.
## Answer: 
At the optimal solution constraint1, 2 and 3 (ie.$x_1+x_2^2+x_3^3-2-3\sqrt2=0,x_2-x_3^2+x_4+2-2\sqrt2=0 \ and \ 
x_1x_5-2=0 $ ) are active since value of RHS is zero in constraints and in optimal solution value of these constrains are approximately zero.So all the three constraints are active constraints. \\
value of $x_3= 1.4728179315306016 $ which does not belongs to set {-5,5} so constraints 4,5 ($-5\leq x_3 \ and  \ x_3 \leq5$)  are inactive. \\
Similarly value of $x_5= 1.6790814362660684 $ which does not belongs to set {0,3} so constraints 6,7 ($0\leq x_5 \ and  \ x_5 \leq3$)  are inactive.



##  6. [R] Solve the model using the following choices of starting points: $(x_1, x_2, x_3, x_4, x_5) =(2^i, 2^i, 2^i, 2^i, 2^i)$ for i = 2, 4, 8, 10, 12. Tabulate the optimal solution, optimal values of optimizationvariables and time taken to solve, for each i. Comment on your observations
## Answer:
First we solve the problem for each i:

In [16]:
a=[2**i for i in [2,4,8,10,12]]

In [17]:
a

[4, 16, 256, 1024, 4096]

In [18]:
df=pd.DataFrame()

In [19]:
df['index']=['optimal Value','x1','x2','x3','x4','x5','time']

In [20]:
df=df.set_index('index')

In [21]:
df

Empty DataFrame
Columns: []
Index: [optimal Value, x1, x2, x3, x4, x5, time]

In [22]:
def diffinit(b):
  for i in b:
    model=ConcreteModel()
    model.x1 = Var(initialize=i)
    model.x2 = Var(initialize=i)
    model.x3 = Var(initialize=i)
    model.x4 = Var(initialize=i)
    model.x5 = Var(initialize=i)
    model.constr1= Constraint(expr= model.x1+(model.x2)**2 + (model.x3)**3 -2 -3*(2)**0.5==0)
    model.constr2= Constraint(expr= model.x2-(model.x3)**2 +model.x4+ 2-2*(2)**0.5==0)
    model.constr3= Constraint(expr= model.x1*model.x5-2==0)
    model.constr4= Constraint(expr= model.x1>=-5)
    model.constr5= Constraint(expr= model.x1<=5)
    model.constr6= Constraint(expr= model.x3>=0)
    model.constr7= Constraint(expr= model.x3<=3)
    model.cost= Objective(rule=fun, sense=minimize)
   
    results = solver.solve(model)
    
    print(f'When initializer of [x1,x2,x3,x4,x5] is {[i,i,i,i,i]}')
    print(results)
    print('\nCost = ', model.cost())

    print('\nOptimal solution:')
    print('x1 = ', model.x1())
    print('x2 = ', model.x2())
    print('x3 = ', model.x3())
    print('x4 = ', model.x4())
    print('x5 = ', model.x5())

    print('\nConstraints')
    print('Constraint 1  = ', model.constr1())
    print('Constraint 2  = ', model.constr2())
    print('Constraint 3  = ', model.constr3())
    print('Constraint 4  = ', model.constr4())
    print('Constraint 5  = ', model.constr5())
    print('Constraint 6  = ', model.constr6())
    print('Constraint 7  = ', model.constr7())
    df[f'2**i={i}']=[model.cost(),model.x1(),model.x2(),model.x3(),model.x4(),model.x5(),results.solver.time]
    print('_'*50)



In [23]:
diffinit(a)

When initializer of [x1,x2,x3,x4,x5] is [4, 4, 4, 4, 4]

Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 5
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.12.13\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.016002178192138672
Solution: 
- number of solutions: 0
  number of solutions displayed: 0


Cost =  0.07877682087105696

Optimal solution:
x1 =  1.1911274562403504
x2 =  1.3626031649434764
x3 =  1.4728179315306016
x4 =  1.6350166192407931
x5 =  1.6790814362660684

Constraints
Constraint 1  =  8.881784197001252e-16
Constraint 2  =  -4.440892098500626e-16
Constraint 3  =  -3.774758283725532e-15
Constraint 4  =  1.1911274562403504
Constraint 5  =  1.1911274562403504
Constraint 6  =  1.4728179315306016
Constraint 7  =  1.4728179315306016
__________________________________________________
When initializer of [x1,x2,x3,x4,x5] is [16, 16, 16, 16, 16]

Problem

In [24]:
Table=df

In [25]:
Table

2**i=4   2**i=16  2**i=256  2**i=1024  2**i=4096
index                                                            
optimal Value  0.078777  0.078777  0.078777   0.078777   0.078777
x1             1.191127  1.191127  1.191127   1.191127   1.191128
x2             1.362603  1.362603  1.362603   1.362603   1.362603
x3             1.472818  1.472818  1.472818   1.472818   1.472818
x4             1.635017  1.635017  1.635017   1.635017   1.635017
x5             1.679081  1.679081  1.679081   1.679081   1.679081
time           0.016002  0.032633  0.030225   0.041327   0.071933

## Answer: 
Above table shows the optimal values of objective function,decision variables,tame taken to solve problem for various initial values of decision variables. \\
Here we can observe that optimized value of objective function and decision variables doesnot change with respect to change in initial value of decision variables.But value of time taken to solve problem changes with respect to change in initial value of decision variables.
By above observation we can say that when the initial values of decision variables are (4,4,4,4,4) then time taken to solve the problem is least time.As soon as initial value increases time taken to solve the problem is also increases.Since we got the optimal values of decision variables are : \\

x1 =  1.1911275027451427 \\
x2 =  1.3626031771640859 \\
x3 =  1.4728179218935038 \\
x4 =  1.6350165752432628 \\
x5 =  1.6790813649752658 \\

So when initial value, near to optimal values it took least time time to solve the problem.as soon as difference  between initial value and optimal value increases time taken to solve problem also increases.